src/lib/x509/x509_builder.cpp

```diff
@@ -0,0 +1,288 @@
+         if(m_is_ca_request) {
+            throw Invalid_State("CertificateParametersBuilder::set_as_ca_certificate cannot be called twice");
+         } else {
+            m_is_ca_request = true;
+            m_path_limit = path_limit;
+         }
+      }
+
+   private:
+      void add_subject_dn(std::string_view fn_suffix, std::string_view attr, std::string_view value) {
+         const auto oid = OID::from_string(attr);
+         const size_t ub = X509_DN::lookup_ub(oid);
+
+         if(value.empty()) {
+            throw Invalid_Argument(fmt("CertificateParametersBuilder::add_{}: empty name is prohibited", fn_suffix));
+         }
+
+         if(value.size() > ub) {
+            throw Invalid_Argument(
+               fmt("CertificateParametersBuilder::add_{}: name exceeds maximum allowed length ({}) for this type",
+                   fn_suffix,
+                   ub));
+         }
+         m_subject_dn.add_attribute(oid, value);
+      }
+
+      std::unique_ptr<Certificate_Extension> setup_alt_name(const Extensions& extensions) const {
+         AlternativeName subject_alt;
+
+         /*
+         If the extension was already created in extensions we need to merge the
+         values provided with the extension value
+         */
+         if(const auto* ext = extensions.get_extension_object_as<Cert_Extension::Subject_Alternative_Name>()) {
+            subject_alt = ext->get_alt_name();
+         }
+
+         for(const auto& dns : m_dns) {
+            subject_alt.add_dns(dns);
+         }
+         for(const auto& uri : m_uri) {
+            subject_alt.add_uri(uri);
+         }
+         for(const auto& email : m_email) {
+            subject_alt.add_email(email);
+         }
+         for(const auto& xmpp : m_xmpp) {
+            subject_alt.add_other_name(OID::from_string("PKIX.XMPPAddr"), ASN1_String(xmpp, ASN1_Type::Utf8String));
+         }
+         for(const uint32_t ipv4 : m_ipv4) {
+            subject_alt.add_ipv4_address(ipv4);
+         }
+
+         return std::make_unique<Cert_Extension::Subject_Alternative_Name>(subject_alt);
+      }
+
+      X509_DN m_subject_dn;
+      Extensions m_extensions;
+      Key_Constraints m_usage;
+      std::vector<std::string> m_email;
+      std::vector<std::string> m_dns;
+      std::vector<std::string> m_uri;
+      std::vector<std::string> m_xmpp;
+      std::vector<uint32_t> m_ipv4;
+      std::vector<OID> m_ext_usage;
+      bool m_is_ca_request = false;
+      std::optional<size_t> m_path_limit;
+};
+
+CertificateParametersBuilder::CertificateParametersBuilder() :
+      m_state(std::make_unique<CertificateParametersBuilder::State>()) {}
+
+CertificateParametersBuilder::CertificateParametersBuilder(CertificateParametersBuilder&& other) noexcept = default;
+
+CertificateParametersBuilder::~CertificateParametersBuilder() = default;
+
+X509_Certificate CertificateParametersBuilder::into_self_signed_cert(std::chrono::system_clock::time_point not_before,
```

Comment:
For the PKCS#10 req I can add a serial number when signing, here I can't. Would it maybe be possible to add the serial no to the builder itself, so that I don't have to duplicate that functionality?